### DISCUSSION TITLE
nexts/script caching script

### DISCUSSION THREAD

**Discussion by D-Thrane** [vercel/next.js#72022] [Help]
### Summary

So we have a webmaster that has a JS Block in our CMS where they can load some JavaScript. So they have some dynamic 365 forms with is loaded with nexts/script and the src comes from the CMS content. The form renders fine on first page load. If you navigate to another page and back again, it does not render det form. Looks like it caches the script and does not load it again. And it is needed as the dynamic 365 script renders some html inside some content divs with a specific ID. Is there some way to make that work or clear the cache for the route?

**Comment by icyJoseph** [comment]
Hi,

The problem is that the, the script runs once, and that's it. Imagine if you had a next/script, that loaded jquery, or something else, you wouldn't want it to rerun on every navigation. That's a hand wavy explanation, but the idea is that the script is already run. 

Is it possible for you to add some query or with a random value to the script src? so that it is forced to re-add it to the DOM?

**Reply by D-Thrane** [reply]
Did try to add a timestamp to the script src, no luck. Found out that when I use <Link to navigate between pages it does not work. But using <a it works and loads the script every time.

**Reply by icyJoseph** [reply]
> Found out that when I use <Link to navigate between pages it does not work. But using <a it works and loads the script every time.

Because you opt-out from SPA navigation into unloading and reloading the page whole. That's not necessarily a good option.

I wonder if the script you are working with, is doing some checks to avoid re-running. 

Maybe you'd have to manually append/remove the script within a use-effect.

**Reply by D-Thrane** [reply]
I have tried to remove the script on unmount through the useEffect. Does not work even when I can se that the script is removed from the DOM.

**Reply by icyJoseph** [reply]
Show us how you do that. I am starting to suspect that the script might do a window check and bail out from re-running. Classic among this kind of script...

**Reply by D-Thrane** [reply]
Code bellow

**Reply by icyJoseph** [reply]
Yeah already answered. First drop the next/script, that one is primarily optimized for one time loading. 

I'd hope that the query trick, or generating a dynamically changing id would work, but as far as I remember it doesn't necessarily.

**Comment by aniruddhaadak80** [comment]
Hey @D-Thrane ,

To fix the caching issue with your JavaScript not rendering on page navigation, try these:

1. **Disable Caching**: Append a unique query string to the script's `src` (e.g., `your-script-source.js?v=${new Date().getTime()}`).

2. **Reload Script**: Remove the existing script and load it again when navigating:
   ```javascript
   function reloadScript() {
       const existingScript = document.querySelector('script[src="your-script-source.js"]');
       if (existingScript) {
           existingScript.remove();
       }
       const script = document.createElement('script');
       script.src = 'your-script-source.js';
       document.head.appendChild(script);
   }
   ```

Implementing either of these should help ensure the script reloads properly.

Thanks.

**Reply by D-Thrane** [reply]
Here is a test script. I see that it adds d365mkt and d365mktforms to the window object. But setting them to undefined still does not work.

```
export const JSBlock = ({ javascript }: IJSBlock) => {
  // javascript is https://cxppusa1formui01cdnsa01-endpoint.azureedge.net/eur/FormLoader/FormLoader.bundle.js
  const [scriptElement, setScriptElement] = useState<string | undefined>(undefined);

  const timeStamp = new Date().getTime();

  useEffect(() => {
    let doc;
    if (javascript) {
      const parser = new DOMParser();
      doc = parser.parseFromString(javascript, "text/html");
      const htmlScriptElement = doc.querySelector("script");
      const scriptSrc = htmlScriptElement?.getAttribute("src");

      if (scriptSrc) {
        setScriptElement(scriptSrc);
      }
    }

    // Cleanup function to remove the script element on unmount
    return () => {
      if (scriptElement) {
        const existingScript = document.querySelector(`script[src="${scriptElement}?v=${timeStamp}"]`);
        if (existingScript) {
          existingScript.remove();
        }
      }
      setScriptElement(undefined);
      doc = undefined;
    };
  }, [javascript, scriptElement]);

  return (
    <>
      <div
        data-form-id="c01e5449-eccb-ee11-9078-00224885f417"
        data-form-api-url="https://public-eur.mkt.dynamics.com/api/v1.0/orgs/ee12437d-8415-444b-a80c-dbe355805dc0/landingpageforms"
        data-cached-form-url="https://assets-eur.mkt.dynamics.com/ee12437d-8415-444b-a80c-dbe355805dc0/digitalassets/forms/c01e5449-eccb-ee11-9078-00224885f417"
      ></div>

      {scriptElement && (
        <Script
          src={`${scriptElement}?v=${timeStamp}`}
          id="javascript"
          data-culture="EN"
          type="text/javascript"
          strategy="afterInteractive"
        />
      )}
    </>
  );
};
```

**Reply by icyJoseph** [reply]
I think there was a misunderstanding here. Skip the next/script, and rather, just inject a script tag yourself, within the effect. That ought to work.

**Reply by D-Thrane** [reply]
Still the same problem :(

```
export const JSBlock = ({ javascript }: IJSBlock) => {
  const [scriptElement, setScriptElement] = useState<string | undefined>(undefined);

  const timeStamp = new Date().getTime();

  useEffect(() => {
    let doc;
    if (javascript) {
      const parser = new DOMParser();
      doc = parser.parseFromString(javascript, "text/html");
      const htmlScriptElement = doc.querySelector("script");
      const scriptSrc = htmlScriptElement?.getAttribute("src");

      if (scriptSrc) {
        setScriptElement(scriptSrc);
        const existingScript = document.querySelector(`script[src="${scriptSrc}?v=${timeStamp}"]`);
        if (existingScript) {
          existingScript.remove();
        }

        const script = document.createElement("script");
        script.src = `${scriptSrc}?v=${timeStamp}`;
        document.head.appendChild(script);
      }
    }

    // Cleanup function to remove the script element on unmount
    return () => {
      if (scriptElement) {
        const existingScript = document.querySelector(`script[src="${scriptElement}?v=${timeStamp}"]`);
        if (existingScript) {
          existingScript.remove();
        }
      }
      setScriptElement(undefined);
      doc = undefined;
    };
  }, [javascript, scriptElement]);
};

```

**Reply by icyJoseph** [reply]
I'll look into details of your implementation later tonight, but assuming it's correct, then we can start to think that your script is doing something to prevent itself from re-running. Do you know if there are docs for the exposed API of this script?

**Reply by D-Thrane** [reply]
If i link to my own script that does a console.log it works when i re-enter the page. So it must be something inside https://cxppusa1formui01cdnsa01-endpoint.azureedge.net/eur/FormLoader/FormLoader.bundle.js that makes it not want to re-run.

**Reply by icyJoseph** [reply]
Is that the script you are loading? (Not nearby my workstation) 

Usually there's an IIFE wrapper that checks the window object for some trace. Or perhaps a fuse kind of variable, set once and forgotten. I'll try to take a look later.

**Reply by icyJoseph** [reply]
You could also, copy the script into your public folder, add a debugger statement, load it from there instead, and step through to see what's not re-running exactly.

**Reply by D-Thrane** [reply]
Yes, that is the script. There is a if where it does some checks on the window object. I can only set the objects to null or undefined. But still does not work :/

**Reply by icyJoseph** [reply]
Do you know, where does `d365mktConfigure` come from?

**Comment by siemen-infi** [comment]
Hi,

I came across this thread when I was debugging the exact same thing.

- We have our page content in contentful (CMS)
- We have a content model representing the dynamics form, which basically are the props of the component below.
- If by chance you're on react 18, see this approach instead, it's less hacky, the formloader bundle contains some react helper that you can use to get rid of the trick where you generate a new div and swap it by yourself (we couldnt use it because we're on next15 \w react19 and the FormPlaceholder component the example adds is on react18) https://learn.microsoft.com/en-us/dynamics365/customer-insights/journeys/developer/realtime-marketing-form-client-side-extensibility#injecting-a-marketing-form-into-a-react-application

```typescript
export const DynamicsForm: React.FC<DynamicsFormProps> = ({
  formId,
  formApiUrl,
  cachedFormUrl,
  scriptSrc,
}) => {
  const wrapperRef = React.useRef<HTMLDivElement>(null);

  return (
    <>
        <div
          ref={wrapperRef}
          data-form-id={formId}
          data-form-api-url={formApiUrl}
          data-cached-form-url={cachedFormUrl}
        ></div>
      <Script
        src={scriptSrc}
        onReady={() => {
          // There are some issues with the embedded form loader script where it seems that
          //  there's something inside of it that prevents the script from re-running when a user navigates back
          //  to the page where a form is embedded.
          // This bit below manually triggers the correct thing to render a new div with said form and swaps it for the previous one.
          if (wrapperRef?.current?.children.length === 0) {
            const oldDiv = document.querySelector(
              `div[data-form-id='${formId}']`,
            );

            if (!oldDiv) {
              return;
            }

            // @ts-expect-error - d365mktforms is a global variable which gets injected by the ms dynamics formloader script
            // eslint-disable-next-line no-undef
            const newDiv = d365mktforms!.createForm(
              formId,
              formApiUrl,
              cachedFormUrl,
            );

            const parentWrapper = oldDiv?.parentNode;
            parentWrapper?.appendChild(newDiv);
            parentWrapper?.removeChild(oldDiv);
          }
        }}
      />
    </>
  );
};
```

Hope this finds you in good spirit, best wishes!
